File: examples/dns/src/dns.js
import dns from 'node:dns';
import { promises as dnsPromises } from 'node:dns';

export async function test() {
    const results = {};

    // Test 1: Error codes exist
    results.hasErrorCodes = (
        dns.NODATA === 'ENODATA' &&
        dns.NOTFOUND === 'ENOTFOUND' &&
        dns.SERVFAIL === 'ESERVFAIL' &&
        dns.NOTIMP === 'ENOTIMP' &&
        dns.TIMEOUT === 'ETIMEOUT'
    );

    // Test 2: Functions exist
    results.hasLookup = typeof dns.lookup === 'function';
    results.hasResolve = typeof dns.resolve === 'function';
    results.hasResolve4 = typeof dns.resolve4 === 'function';
    results.hasResolve6 = typeof dns.resolve6 === 'function';
    results.hasReverse = typeof dns.reverse === 'function';
    results.hasSetServers = typeof dns.setServers === 'function';
    results.hasGetServers = typeof dns.getServers === 'function';
    results.hasSetDefaultResultOrder = typeof dns.setDefaultResultOrder === 'function';
    results.hasGetDefaultResultOrder = typeof dns.getDefaultResultOrder === 'function';

    // Test 3: Resolver class exists
    results.hasResolver = typeof dns.Resolver === 'function';
    const resolver = new dns.Resolver();
    results.resolverHasResolve = typeof resolver.resolve === 'function';
    results.resolverHasLookup = typeof resolver.lookup === 'function';

    // Test 4: Promises API exists
    results.hasPromises = typeof dns.promises === 'object';
    results.promisesHasLookup = typeof dns.promises.lookup === 'function';
    results.promisesHasResolve = typeof dns.promises.resolve === 'function';
    results.promisesHasResolve4 = typeof dns.promises.resolve4 === 'function';
    results.promisesHasResolve6 = typeof dns.promises.resolve6 === 'function';

    // Test 5: setDefaultResultOrder / getDefaultResultOrder
    dns.setDefaultResultOrder('ipv4first');
    results.resultOrderSet = dns.getDefaultResultOrder() === 'ipv4first';
    dns.setDefaultResultOrder('verbatim');

    // Test 6: setServers / getServers
    dns.setServers(['8.8.8.8', '8.8.4.4']);
    const servers = dns.getServers();
    results.serversSet = servers.length === 2 && servers[0] === '8.8.8.8';

    // Test 7: lookup with IP address passthrough (should work without network)
    results.ipPassthrough = await new Promise((resolve) => {
        dns.lookup('127.0.0.1', (err, address, family) => {
            resolve(!err && address === '127.0.0.1' && family === 4);
        });
    });

    // Test 8: lookup with IP passthrough (IPv6)
    results.ipv6Passthrough = await new Promise((resolve) => {
        dns.lookup('::1', (err, address, family) => {
            resolve(!err && address === '::1' && family === 6);
        });
    });

    // Test 9: lookup with null hostname returns loopback
    results.nullHostname = await new Promise((resolve) => {
        dns.lookup('', (err, address, family) => {
            resolve(!err && address === '127.0.0.1' && family === 4);
        });
    });

    // Test 10: lookup with all: true and IP passthrough
    results.lookupAll = await new Promise((resolve) => {
        dns.lookup('127.0.0.1', { all: true }, (err, addresses) => {
            resolve(!err && Array.isArray(addresses) && addresses.length === 1 &&
                addresses[0].address === '127.0.0.1' && addresses[0].family === 4);
        });
    });

    // Test 11: Promises lookup with IP passthrough
    try {
        const promiseResult = await dnsPromises.lookup('127.0.0.1');
        results.promiseLookup = promiseResult.address === '127.0.0.1' && promiseResult.family === 4;
    } catch (e) {
        results.promiseLookup = false;
    }

    // Test 12: lookup with family mismatch should error
    results.familyMismatch = await new Promise((resolve) => {
        dns.lookup('127.0.0.1', { family: 6 }, (err) => {
            resolve(err && err.code === 'ENOTFOUND');
        });
    });

    // Test 13: unsupported record types return ENOTIMP
    results.unsupportedMx = await new Promise((resolve) => {
        dns.resolveMx('example.com', (err) => {
            resolve(err && err.code === 'ENOTIMP');
        });
    });

    // Test 14: Hint flags exist
    results.hasAddrconfig = dns.ADDRCONFIG === 1024;
    results.hasV4mapped = dns.V4MAPPED === 8;
    results.hasAll = dns.ALL === 16;

    return JSON.stringify(results);
}

File: tests/runtime/dns.rs
use crate::common::{CompiledTest, invoke_and_capture_output};
use test_r::{inherit_test_dep, test};
use wasmtime::component::Val;

inherit_test_dep!(
    #[tagged_as("dns")]
    CompiledTest
);

#[test]
async fn dns_module_api(#[tagged_as("dns")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (result, _output) =
        invoke_and_capture_output(compiled.wasm_path(), None, "test", &[]).await;
    let result = result?;

    if let Some(Val::String(json_str)) = result {
        let result_obj: serde_json::Value = serde_json::from_str(&json_str)?;

        // Error codes
        assert!(
            result_obj["hasErrorCodes"].as_bool().unwrap_or(false),
            "dns error codes should exist"
        );

        // Functions exist
        assert!(
            result_obj["hasLookup"].as_bool().unwrap_or(false),
            "dns.lookup should exist"
        );
        assert!(
            result_obj["hasResolve"].as_bool().unwrap_or(false),
            "dns.resolve should exist"
        );
        assert!(
            result_obj["hasResolve4"].as_bool().unwrap_or(false),
            "dns.resolve4 should exist"
        );
        assert!(
            result_obj["hasResolve6"].as_bool().unwrap_or(false),
            "dns.resolve6 should exist"
        );
        assert!(
            result_obj["hasReverse"].as_bool().unwrap_or(false),
            "dns.reverse should exist"
        );
        assert!(
            result_obj["hasSetServers"].as_bool().unwrap_or(false),
            "dns.setServers should exist"
        );
        assert!(
            result_obj["hasGetServers"].as_bool().unwrap_or(false),
            "dns.getServers should exist"
        );
        assert!(
            result_obj["hasSetDefaultResultOrder"]
                .as_bool()
                .unwrap_or(false),
            "dns.setDefaultResultOrder should exist"
        );
        assert!(
            result_obj["hasGetDefaultResultOrder"]
                .as_bool()
                .unwrap_or(false),
            "dns.getDefaultResultOrder should exist"
        );

        // Resolver class
        assert!(
            result_obj["hasResolver"].as_bool().unwrap_or(false),
            "dns.Resolver should exist"
        );
        assert!(
            result_obj["resolverHasResolve"].as_bool().unwrap_or(false),
            "Resolver.resolve should exist"
        );
        assert!(
            result_obj["resolverHasLookup"].as_bool().unwrap_or(false),
            "Resolver.lookup should exist"
        );

        // Promises API
        assert!(
            result_obj["hasPromises"].as_bool().unwrap_or(false),
            "dns.promises should exist"
        );
        assert!(
            result_obj["promisesHasLookup"].as_bool().unwrap_or(false),
            "dns.promises.lookup should exist"
        );
        assert!(
            result_obj["promisesHasResolve"].as_bool().unwrap_or(false),
            "dns.promises.resolve should exist"
        );
        assert!(
            result_obj["promisesHasResolve4"].as_bool().unwrap_or(false),
            "dns.promises.resolve4 should exist"
        );
        assert!(
            result_obj["promisesHasResolve6"].as_bool().unwrap_or(false),
            "dns.promises.resolve6 should exist"
        );

        // setDefaultResultOrder / getDefaultResultOrder
        assert!(
            result_obj["resultOrderSet"].as_bool().unwrap_or(false),
            "setDefaultResultOrder should work"
        );

        // setServers / getServers
        assert!(
            result_obj["serversSet"].as_bool().unwrap_or(false),
            "setServers/getServers should work"
        );

        // IP passthrough
        assert!(
            result_obj["ipPassthrough"].as_bool().unwrap_or(false),
            "lookup('127.0.0.1') should pass through"
        );
        assert!(
            result_obj["ipv6Passthrough"].as_bool().unwrap_or(false),
            "lookup('::1') should pass through"
        );
        assert!(
            result_obj["nullHostname"].as_bool().unwrap_or(false),
            "lookup('') should return loopback"
        );
        assert!(
            result_obj["lookupAll"].as_bool().unwrap_or(false),
            "lookup with all:true should work"
        );

        // Promise lookup
        assert!(
            result_obj["promiseLookup"].as_bool().unwrap_or(false),
            "promises.lookup should work with IP"
        );

        // Family mismatch
        assert!(
            result_obj["familyMismatch"].as_bool().unwrap_or(false),
            "family mismatch should return ENOTFOUND"
        );

        // Unsupported record types
        assert!(
            result_obj["unsupportedMx"].as_bool().unwrap_or(false),
            "resolveMx should return ENOTIMP"
        );

        // Hint flags
        assert!(
            result_obj["hasAddrconfig"].as_bool().unwrap_or(false),
            "ADDRCONFIG should be 1024"
        );
        assert!(
            result_obj["hasV4mapped"].as_bool().unwrap_or(false),
            "V4MAPPED should be 8"
        );
        assert!(
            result_obj["hasAll"].as_bool().unwrap_or(false),
            "ALL should be 16"
        );
    } else {
        anyhow::bail!("Expected string result from test function");
    }

    Ok(())
}
